006-graphql-pagination.md

```diff
@@ -0,0 +1,84 @@
+# GraphQL Pagination
+
+We should use [Relay Cursor Connections](https://facebook.github.io/relay/graphql/connections.htm#) to paginate results in GraphQL. 
+
+**Discussion:** https://github.com/DoSomething/rfcs/pull/6
+
+## Problem
+
+We currently return plain lists, like `[Post]`, for paginated GraphQL queries:
+
+```gql
+{
+  posts(count: 10, page: $page) {
+    id
+    type
+    url
+  }
+}
+```
+
+This gives us no way to know whether there's another page of results without making a request to check. Because we built these on top of standard `OFFSET` pagination, there are also some performance issues with larger datasets & the potential for duplicate results to appear (described in more detail under "Alternatives").
+
+## Proposal
+
+We should use [Relay Cursor Connections](https://facebook.github.io/relay/graphql/connections.htm#) to paginate results in GraphQL, and add REST API support to paginate via cursors (e.g. "get me all the results after #1015" rather than "get me the tenth page of results") to Northstar and Rogue.
+
+```gql
+{
+  paginatedPosts(first: 10, after: $cursor) {
+    edges {
+      cursor
+      node {
+        id
+        type
+        url
+      }
+    }
+    pageInfo {
+      hasNextPage
+    }
+  }
+}
+```
+
+This relies on generating "opaque" cursors for each element in the results. The cursor, e.g. `Nzk1MS4yNDA3Nw==` encodes information about where the item appears in the results (the ID & the sorted column value, if applicable).
+
+## Drawbacks
```

Comment:
I'm not sure if this drawback is worth explicitly calling out, but in the Shopify article they mention:
> Relative cursor pagination remembers where you were so that each request after the first continues from where the previous request left off. The downside is that you can no longer jump to a specific page.

Though I doubt that's a big deal for us in practice?

Reply:
Oh, I definitely think that's worth calling out! I'll add that in here.

Reply:
I do think this is a good flag, however I think in the implementations we've had recently we are moving away from the `next` & `previous` page and more towards the infinite scroll, so it could be a non issue.

Reply:
I don't think this update ever got added in?

Reply:
Oops! Added in 9cc2dae.